### PR TITLE
Начнём с конца

### PR DESCRIPTION
**Pull request by oblasov**
---
:mortar_board: [Начнём с конца](https://up.htmlacademy.ru/ecmascript/3/user/124494/tasks/6)

:boom: https://htmlacademy-ecmascript.github.io/124494-pixel-hunter/5/

### CODE REVIEW

File: js/data/game-data.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+/**
+ * Стартовые параметры игры
+ * @enum {number}
+ */
+export const startConfigs = {
```
Comment (balesniy) [inline js/data/game-data.js]: enum просят называть в ед числе и сами поля заглавными, посмотри критерий Б15

File: js/data/game-data.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Функция подсчета набранных пользователем очков
+ * @param {Array.<Object>} userAnswers
+ * @param {number} userLives
+ * @return {number}
+ */
+export const countUserScore = (userAnswers, userLives) => {
+  let score = 0;
+
+  if (typeof userAnswers !== `object` || userAnswers === null || typeof userLives !== `number`) {
```
Comment (balesniy) [inline js/data/game-data.js]: userAnswers заявлен массивом, можно проверить явно на массив, `Array.isArray()`. и мне кажется, что лучше бросить ошибку в случае неверных параметров.

File: js/data/game-data.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    return null;
+  }
+
+  if (userAnswers.length < 10 || userLives < 0) {
+    return -1;
+  }
+
+  // Производим подсчет заработанных очков
+  userAnswers.forEach((answer) => {
```
Comment (balesniy) [inline js/data/game-data.js]: чтобы свернуть массив в число лучше воспользоваться reduce

File: js/data/game-data.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    // За каждый правильный ответ
+    if (answer.correct) {
+      // даётся 100 очков.
+      score += 100;
+      switch (answer.speed) {
+        // За каждый быстрый ответ дополнительно начисляется 50 очков
+        // Таким образом, быстрый ответ приносит игроку 150 очков.
+        case `fast`:
```
Comment (balesniy) [inline js/data/game-data.js]: значения типов ответов и количества очков просят вынести в константы в enum, посмотри критерий Д17. если они будут в перечислении лежать, то и switch будет не нужен, можно будет значение сразу взять по ключу

File: js/data/game-data.test.js [hunk 5]
```diff
@@ -1,9 +1,92 @@
...
-describe(`Array`, () => {
-  describe(`#indexOf()`, () => {
-    it(`should return -1 when the value is not present`, () => {
-      assert.equal(-1, [1, 2, 3].indexOf(4));
+import {startConfigs, countUserScore} from '../data/game-data.js';
+
+describe(`gameData`, () => {
+  describe(`getUserAnswers`, () => {
+
+    const normalAnswers = [
```
Comment (balesniy) [inline js/data/game-data.test.js]: можно на такой случай использовать `Array.fill()`

File: js/data/game-data.test.js [hunk 6]
```diff
@@ -1,9 +1,92 @@
...
+
+    it(`should return 350, if user gave all answers with slow speed and lost all lives`, () => {
+      assert.strictEqual(countUserScore(badAnswers, 0), 350);
     });
+
+    it(`should return 1650, if user gave all answers with fast speed and saved all lives`, () => {
+      assert.strictEqual(countUserScore(bestAnswers, 3), 1650);
+    });
+
+    it(`should deal with corner cases correctly`, () => {
```
Comment (balesniy) [inline js/data/game-data.test.js]: если такое тестировать, то нужно все возможные случаи перебрать, например `countUserScore({}, startConfigs.lives)` или `countUserScore(normalAnswers, NaN)`.

File: js/game/game.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
+/**
+ * @param {number} value
+ * @constructor
+ */
+export const Timer = function (value = 0) {
```
Comment (balesniy) [inline js/game/game.js]: просят использовать синтаксис классов, посмотри критерий Д5 и Б31

File: js/game/game.test.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    it(`should allow to enter valid data`, () => {
+      assert(typeof (new Timer(startConfigs.timer).value) === `number`);
+    });
+
+    it(`should set time, during which the timer will work`, () => {
+      assert.strictEqual((new Timer(10).value), 10);
+    });
+
+    it(`method tick() decreases the timer by one`, () => {
+      assert.strictEqual((new Timer(startConfigs.timer).tick().value), new Timer(startConfigs.timer).value - 1);
```
Comment (balesniy) [inline js/game/game.test.js]: ты тут сравниваешь два разных таймера
